test/control-plane/services/runner.go

```diff
@@ -0,0 +1,463 @@
+	// Each step currently needs to describe all objects that exist
+	// at that point in time.
+	objCache := newK8sObjectCache()
+
+	datapath, agentHandle := startCiliumAgent(testCase.NodeName, clients, modConfig)
+	defer tearDown(agentHandle)
+
+	lbmap := datapath.LBMap().(*mockmaps.LBMockMap)
+
+	// Run through test steps and validate
+	for _, step := range testCase.Steps {
+		updated, deleted := objCache.updateObjects(step.Inputs)
+
+		for _, obj := range updated {
+			gvr, ns, name := gvrAndName(obj)
+
+			updateOrAdd := func(obj k8sRuntime.Object, tracker k8sTesting.ObjectTracker) {
+				if _, err := tracker.Get(gvr, ns, name); err == nil {
+					if err := tracker.Update(gvr, obj, ns); err != nil {
+						panic(err)
+					}
+				} else {
+					if err := tracker.Add(obj); err != nil {
+						panic(err)
+					}
+				}
+			}
+
+			if coreDecoder.known(obj) {
+				updateOrAdd(coreDecoder.convert(obj), coreTracker)
+			}
+			if slimDecoder.known(obj) {
+				updateOrAdd(slimDecoder.convert(obj), slimTracker)
+			}
+			if ciliumDecoder.known(obj) {
+				updateOrAdd(ciliumDecoder.convert(obj), ciliumTracker)
+			}
+
+		}
+
+		for _, obj := range deleted {
+			gvr, ns, name := gvrAndName(obj)
+			if coreDecoder.known(obj) {
+				if err := coreTracker.Delete(gvr, ns, name); err != nil {
+					panic(err)
+				}
+			}
+			if slimDecoder.known(obj) {
+				if err := slimTracker.Delete(gvr, ns, name); err != nil {
+					panic(err)
+				}
+			}
+			if ciliumDecoder.known(obj) {
+				if err := ciliumTracker.Delete(gvr, ns, name); err != nil {
+					panic(err)
+				}
+			}
+		}
+
+		// Validate the datapath state. Since the processing of the k8s objects is asynchronous
+		// and there is no obvious way to synchronize with datapath (yet), so we'll
+		// try a few times and wait a bit.
+		err := retryForDuration(
+			func() error { return step.Validate(lbmap) },
+			10*time.Second)
+
+		// Check that the state stays correct and consistent over a short period of time
+		// after the initial match.
+		for retries := 3; err == nil && retries > 0; retries-- {
+			time.Sleep(100 * time.Millisecond)
+			err = step.Validate(lbmap)
+		}
+		if err != nil {
+			t.Fatalf("Test failed: %s", err)
+		}
+	}
+}
+
+// NewGoldenTest creates a test suite from YAML files defining
+// the input states and expected LBMap output.
+//
+// The input states are specified in files state<N>.yaml.
+// These are expected to be a k8s "List", e.g. the format of "kubectl get <res> -o yaml".
+// They're fed to the control-plane in lexicographical order and after each the
+// lbmap state is validated against the expected state as described by lbmap<N>.golden.
+func NewGoldenTest(t *testing.T, nodeName string) (testCase *ServicesTestCase) {
+	testCase = &ServicesTestCase{}
+	testCase.NodeName = nodeName
+
+	if _, err := os.Stat("init.yaml"); err == nil {
+		bs, err := os.ReadFile("init.yaml")
+		if err != nil {
+			t.Fatal(err)
+		}
+		testCase.InitialObjects = unmarshalList(bs)
+	}
+
```

Comment:
What happens if `os.Stat` fails?

Reply:
Then the set of initial objects is empty, which will cause the agent to fail to start as it can't find the `Node` resource for the local node. This wasn't the case in earlier iteration, but now that it is, I'll change this to fail here.